### PR TITLE
Fix Code Trimmed Issue

### PR DESCRIPTION
**Pull request by devKshitijJain**
The issue with [Moanco](https://microsoft.github.io/monaco-editor/) is that it loads up the stylesheet when the code editor is visible and then our MHTML format is not able to capture code which is scrollable inside that. To fix this, we can reduce the size of chromium to just 1px to always ensure that the code block will always be down of current visible area. Also added logic to close the page once the screenshot, pdf or html file is generated.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -204,8 +204,8 @@ async function downloadPage(title: string, link: string, doToggleMenu: boolean):
     await setTimeoutPromise(300); // Wait for munu to close
 
     await page.screenshot({ path: `${SAVE_DESTINATION}/${normalizedTitle}.png`, fullPage: true });
-    await page.close();
   }
+  await page.close();
```
Comment (shihabmridha) [inline src/app.ts:208]: Good catch!

**Review by shihabmridha — APPROVED** [review]

File: src/app.ts [hunk 2]
```diff
@@ -41,7 +41,7 @@ async function main(): Promise<void> {
     return;
   }
 
-  browser = await launch({ userDataDir: ROOT_PATH + '/data', headless: true });
+  browser = await launch({ userDataDir: ROOT_PATH + '/data', headless: true, defaultViewport: null, args: ['--window-size=1,1'] });
```
Comment (devKshitijJain) [inline src/app.ts:44]: When headless is true, it's not getting logged in if not already login. But if I change it to the false,  it works. Can you please take a look into this?

Comment (devKshitijJain) [reply src/app.ts:44]: @shihabmridha

Comment (shihabmridha) [reply src/app.ts:44]: Will take a look. Thanks.

Comment (shihabmridha) [reply src/app.ts:44]: **Strange!** I did not encounter any issue. Removed the **data** directory to remove session etc. Then ran the script with `loginCheck: true`. Got logged in and downloaded the course without any issue.

Comment (devKshitijJain) [reply src/app.ts:44]: I am not sure but for me it's not getting logged in. Let me share the error it throws

Comment (devKshitijJain) [reply src/app.ts:44]: `⠹ Loggin in.Node is either not visible or not an HTMLElement
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! educative.io-downloader@1.0.0 start: `node build/src/app.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the educative.io-downloader@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/kshitij/.npm/_logs/2020-04-21T14_18_50_409Z-debug.log`

Comment (devKshitijJain) [reply src/app.ts:44]: this is happening only after adding my changes, without it,  it's working fine. But my changes are required in order to capture full code.

Comment (shihabmridha) [reply src/app.ts:44]: A quick solution would be launch a browser without your change (defaultViewport, args) to login. Then close the browser and open a new one with your changes.

I see an old issue related to this (https://github.com/puppeteer/puppeteer/issues/2977) as well.

BTW, Which OS are you using? What is the screen resolution of your display?

Comment (devKshitijJain) [reply src/app.ts:44]: Using mac, yes that would work but I don't want this manual step. Any other way it can be solved?

Comment (devKshitijJain) [reply src/app.ts:44]: How about including this step in script only? I mean when login it will be without my changes and then it will close and launch another browser instance with my changes?

Comment (devKshitijJain) [reply src/app.ts:44]: anywhere we can connect and chat?

Comment (shihabmridha) [reply src/app.ts:44]: My Hangouts: shihabmridha@gmail.com

Comment (shihabmridha) [reply src/app.ts:44]: Discord username: DIJKSTRA#8375
Sent you an invitation to Hangouts as well.

Comment (thomasgeoff) [reply src/app.ts:44]: When tried it is working but downloading almost blank images or pdf or HTML. Do u know what might be the issue?

Comment (devKshitijJain) [reply src/app.ts:44]: Make sure you have HTML selected. Secondly make headless = false and then try. Which course are you trying to download?

Comment (devKshitijJain) [reply src/app.ts:44]: @thomasgeoff the main problem I think is. You are not getting logged in. In config file change loginCheck to true so it will check first if you are logged in or not and then will proceed further.

Comment (thomasgeoff) [reply src/app.ts:44]: @devKshitijJain did that and it logged in and started downloading. I tried all the formats. Images only show a broken centre line, pdf's are blank and HTML have only the first page. I'm trying to download grokking the coding interview patterns. Is it working for you?

Comment (devKshitijJain) [reply src/app.ts:44]: Yes working perfectly for me

**Comment by thomasgeoff** [comment]
@ devKshitijJain Ok then let me try again with HTML format. Thank you for the response.

File: src/app.ts (hunk 2 above)
Comment (thomasgeoff) [reply src/app.ts:44]: @devKshitijJain Thank you for the response. I tried again with HTML format in Windows 10 PC. Im getting this error:
Processing batch of lessons.Evaluation failed: TypeError: Cannot read property 'childNodes' of undefined at __puppeteer_evaluation_script__:4:80

I haven't worked with JS much so I'm not able to debug this. Can you help me out?

Comment (thomasgeoff) [reply src/app.ts:44]: @shihabmridha @devKshitijJain When I try with loginCheck: true. It gives me this error:
> node build/src/app.js

/ Loggin in.Node is either not visible or not an HTMLElement
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! educative.io-downloader@1.0.0 start: `node build/src/app.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the educative.io-downloader@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\tgeof\AppData\Roaming\npm-cache\_logs\2020-04-24T03_59_13_005Z-debug.log

Comment (devKshitijJain) [reply src/app.ts:44]: @thomasgeoff this is known issue. Just make headless=false and then try.

Comment (devKshitijJain) [reply src/app.ts:44]: @thomasgeoff already discussed this with @shihabmridha. Will work on fix.
